common/workunit/workflow.cpp

```diff
@@ -613,12 +776,816 @@ WorkflowMachine::WorkflowMachine(const IContextLogger &_logctx)
+}
+bool WorkflowMachine::checkIfDone()
+{
+    if((pendingContingencies == 0) && (rootReached))
+    {
+        if( (branchCount == 0) || failureClauseOld)
+        {
+        #ifdef TRACE_WORKFLOW
+                LOG(MCworkflow, "workflow done");
+        #endif
+            done = true;
+            itemQueueSem.signal(numThreads);
+            return true;
+        }
+    }
+    return false;
+}
+void WorkflowMachine::processWfItems()
+{
+    Owned<WorkflowException> error;
+    while(!done)
+    {
+        itemQueueSem.wait();
+        if(!done)
+        {
+            unsigned currentWfid = 0;
+            //enter critical section
+            {
+                CriticalBlock thisBlock(queueCritSec);
+                currentWfid = itemQueue.front();
+                itemQueue.pop();
+            }
+            try
+            {
+                executeItemParallel(currentWfid);
+            }
+            catch(WorkflowException * e)
+            {
+                error.setown(e);
+                done = true;
+                break;
+            }
+        }
+    }
+    if(error)
+        throw error.getLink();
+}
+void WorkflowMachine::performParallel(IGlobalCodeContext *_ctx, IEclProcess *_process)
+{
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "starting perform parallel");
+    #endif
+    ctx = _ctx;
+    process = _process;
+
+    //relink workflow
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Starting to mark Items with their successors");
+    #endif
+    addSuccessors();
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Finished marking Items with their successors");
+    #endif
+
+    //add initial values
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Adding initial workflow items");
+    #endif
+    initialiseItemQueue();
+
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Initialising threads");
+    #endif
+
+    //initialise thread count
+    numThreads = getThreadNumFlag();
+    if(numThreads < 1)
+            numThreads = 4;
+    unsigned maxThreads = getAffinityCpus();
+    if(numThreads > maxThreads)
+        numThreads = maxThreads;
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "num threads = %u", numThreads);
+    #endif
+
+    std::vector<std::thread *> threads(numThreads);
+    //loop through and start threads
+    for(int i=0; i < numThreads; i++)
+        threads[i] = new std::thread([this]() {  this->processWfItems(); });
+
+
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Calling join threads");
+    #endif
+    //loop through and call join
+    for(int i=0; i < numThreads; i++)
+        threads[i]->join();
+
+    #ifdef TRACE_WORKFLOW
+            LOG(MCworkflow, "Destroying threads");
+    #endif
+    //loop through and destroy
+    for(int i=0; i < numThreads; i++)
+        delete threads[i];
+}
+bool WorkflowMachine::isParallelViable()
+{
+    //initialise parallel flag from workunit
+    parallel = getParallelFlag();
```

Comment:
This could be changed to 
   if (!getParallelFlag())
      return false;
and remove the parallel member.  Fine to leave though